db/database.go

```diff
@@ -1738,3 +1753,31 @@ func (context *DatabaseContext) IsGuestReadOnly() bool {
 func (context *DatabaseContext) ForceAPIForbiddenErrors() bool {
 	return context.Options.UnsupportedOptions != nil && context.Options.UnsupportedOptions.ForceAPIForbiddenErrors
 }
+
+//////// TIMEOUTS
+
+// Calls a function, synchronously, while imposing a timeout on the Database's Context. Any call to CheckTimeout while the function is running will return an error if the timeout has expired.
+// The function will *not* be aborted automatically! Its code must check for timeouts by calling CheckTimeout periodically, returning once that produces an error.
+func (db *Database) WithTimeout(timeout time.Duration, operation func() error) error {
```

Comment:
I'm OK with not restoring the old Context on the way out, if a Database instance isn't going to outlive the operation that sets a timeout.

Having a separate Context seems error-prone since the Database already has a Context, and one might forget which is the right one to use.

Reply:
That sounds reasonable - let's just leave it as-is on the way out.